src/librustc/traits/structural_impls.rs

```diff
@@ -618,3 +618,186 @@ impl<'tcx> TypeFoldable<'tcx> for traits::ObligationCause<'tcx> {
+        match *self {
+            Holds(wc) => write!(fmt, "{}", wc),
+            WellFormed(Implemented(ref trait_ref)) => write!(fmt, "WellFormed({})", trait_ref),
+            WellFormed(ProjectionEq(ref projection)) => write!(fmt, "WellFormed({})", projection),
+            FromEnv(Implemented(ref trait_ref)) => write!(fmt, "FromEnv({})", trait_ref),
+            FromEnv(ProjectionEq(ref projection)) => write!(fmt, "FromEnv({})", projection),
+            WellFormedTy(ref ty) => write!(fmt, "WellFormed({})", ty),
+            FromEnvTy(ref ty) => write!(fmt, "FromEnv({})", ty),
+            RegionOutlives(ref predicate) => write!(fmt, "RegionOutlives({})", predicate),
+            TypeOutlives(ref predicate) => write!(fmt, "TypeOutlives({})", predicate),
+        }
+    }
+}
+
+impl fmt::Display for traits::QuantifierKind {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use traits::QuantifierKind::*;
+        match *self {
+            Universal => write!(fmt, "forall"),
+            Existential => write!(fmt, "exists"),
+        }
+    }
+}
+
+impl<'tcx> fmt::Display for traits::LeafGoal<'tcx> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use traits::LeafGoal::*;
+        match *self {
+            DomainGoal(ref domain_goal) => write!(fmt, "{}", domain_goal),
+        }
+    }
+}
+
+impl<'tcx> fmt::Display for traits::Goal<'tcx> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use traits::Goal::*;
+        match *self {
+            Implies(ref hypotheses, ref goal) => {
+                write!(fmt, "if (")?;
+                for (index, hyp) in hypotheses.iter().enumerate() {
+                    if index > 0 {
+                        write!(fmt, ", ")?;
+                    }
+                    write!(fmt, "{}", hyp)?;
+                }
+                write!(fmt, ") {{ {} }}", goal)
+            }
+            And(ref goal1, ref goal2) => write!(fmt, "({}, {})", goal1, goal2),
+            Not(ref goal) => write!(fmt, "not {{ {} }}", goal),
+            Leaf(ref goal) => write!(fmt, "{}", goal),
+            Quantified(qkind, ref goal) => {
+                // FIXME: appropriate binder names
+                write!(fmt, "{}<> {{ {} }}", qkind, goal.skip_binder())
+            }
+        }
+    }
+}
+
+impl<'tcx> fmt::Display for traits::ProgramClause<'tcx> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.consequence)?;
+        if self.conditions.is_empty() {
+            write!(fmt, ".")?;
+        } else {
+            write!(fmt, " :- ")?;
+            for (index, condition) in self.conditions.iter().enumerate() {
+                if index > 0 {
+                    write!(fmt, ", ")?;
+                }
+                write!(fmt, "{}", condition)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<'tcx> TypeFoldable<'tcx> for traits::WhereClauseAtom<'tcx> {
```

Comment:
There are pretty decent macros for this... oh, shoot, they are in https://github.com/rust-lang/rust/pull/48411 -- oh, wait, that just landed =)

You may want to adopt one of these:

https://github.com/rust-lang/rust/blob/8c4ff22a2d745097197c659ef9e3b04b8ceeb070/src/librustc/macros.rs#L160-L164